win32/devdoc/socket_transport_win32.md

```diff
@@ -0,0 +1,296 @@
+
+If `sm_open_begin` does not return `SM_EXEC_GRANTED`, `socket_transport_connect` shall fail and return a non-zero value.
+
+`socket_transport_connect` shall call `socket` with the params `AF_INET`, `SOCK_STREAM` and `IPPROTO_TCP`.
+
+`socket_transport_connect` shall call `connect_to_endpoint` to connect to the endpoint.
+
+If successful `socket_transport_connect` shall call `sm_open_end` with `true`.
+
+If any failure is encountered, `socket_transport_connect` shall call `sm_open_end` with `false`, fail and return a non-zero value.
+
+### connect_to_endpoint
+
+```c
+static int connect_to_endpoint(SOCKET client_socket, const ADDRINFO* addrInfo, uint32_t timeout_usec)
+```
+
+`connect_to_endpoint` shall call the `connect` API to asynchronously connect to the endpoint.
+
+If the `connect` call fails, `connect_to_endpoint` shall check to `WSAGetLastError` for `WSAEWOULDBLOCK`.
+
+On `WSAEWOULDBLOCK` `connect_to_endpoint` shall call the `select` API with a timeout set in the `zrpc_tcp_client_io_create` function and check the return value:
+
+- If the return is `SOCKET_ERROR`, this indicates a connect and `connect_to_endpoint` shall fail.
+
+- If the return value is 0, this indicates a timeout and `connect_to_endpoint` shall fail.
+
+- Any other value this indicates a possible success and and `connect_to_endpoint` shall test if the socket is writable by calling `FD_ISSET`.
+
+If the socket is writable `connect_to_endpoint` shall succeed and return a 0 value.
+
+If any error is encountered `connect_to_endpoint` shall return a non-zero value.
+
+### socket_transport_disconnect
+
+```c
+MOCKABLE_FUNCTION(, void, socket_transport_disconnect, SOCKET_TRANSPORT_HANDLE, socket_transport);
+```
+
+`socket_transport_disconnect` shall disconnect a connected socked from its endpoint.
+
+If `socket_transport` is `NULL`, `socket_transport_disconnect` shall fail and return.
+
+`socket_transport_disconnect` shall call `sm_close_begin` to begin the closing process.
+
+If `sm_close_begin` does not return `SM_EXEC_GRANTED`, `socket_transport_disconnect` shall fail and return.
+
+`socket_transport_disconnect` shall call `closesocket` to disconnect the connected socket.
+
+`socket_transport_disconnect` shall call `sm_close_end`.
+
+### socket_transport_send
+
+```c
+MOCKABLE_FUNCTION(, SOCKET_SEND_RESULT, socket_transport_send, SOCKET_TRANSPORT_HANDLE, socket_transport, SOCKET_BUFFER*, payload, uint32_t, buffer_count, uint32_t*, bytes_written, uint32_t, flags, void*, overlapped_data);
+```
+
+If `socket_transport` is `NULL`, `socket_transport_send` shall fail and return `SOCKET_SEND_ERROR`.
+
+If `payload` is `NULL`, `socket_transport_send` shall fail and return `SOCKET_SEND_ERROR`.
+
+If `buffer_count` is `0`, `socket_transport_send` shall fail and return `SOCKET_SEND_ERROR`.
+
+`socket_transport_send` shall call `sm_exec_begin`.
+
+If `sm_exec_begin` does not return `SM_EXEC_GRANTED`, `socket_transport_send` shall fail and return `SOCKET_SEND_ERROR`.
```

Comment:
Can we fail with something like SOCKET_CLOSED to be explicit here with the difference between "something failed to send" and "we didn't even try because we know the socket is not connected"?

Reply:
Well, I guess we have SOCKET_SEND_FAILED below to differentiate, but still seems that a "NOT_OPEN" or "CLOSED" value may be useful for layers above (vs. invalid args above this).

Reply:
Will do.